proposals/1141_audit-logging/README.md

```diff
@@ -0,0 +1,323 @@
+may cause more than one event to be logged. If possible, unique ID will be determined for each incoming request, and
+will be included in all events caused by that request.
+
+Where possible, the top-level keys of the JSON object will use standardized names. Other top-level keys specific to that
+action type may be added. All keys should be included in documentation for the audit log feature.
+
+Every event should include these keys:
+
+- `time`: the timestamp of the event
+- `event`: the event type, which is a brief description of what happened, with no string interpolation, so it will
+  always be the same for a given event type (e.g. `upstream refresh succeeded`)
+- `v`: a number specifying the format version of the event type, starting with `1`, to give us flexibility to make
+  breaking changes to the format of an event type in future releases (e.g. change the name of the JSON keys, or change
+  the data type of the value of an existing key)
+
+Depending on the event type, an event might include other keys, such as:
+
+- `msg`: a freeform warning or error message meant to be read by a human (e.g. the error message that was returned by an
+  upstream IDP during a failed login attempt)
+- `requestID`: a unique ID for the request, if the event is related to an API request
+- `requestPath`: the path of the endpoint, if the event is related to an API request
+- `requestorIP`: the client's IP, if the event is related to an API request
+- `user`: the username of the user performing the action, if there is one
+- `groups`: the group memberships of the user performing the action, if the action is related determining or changing
+  their group memberships
+
+The details of these additional keys will be worked out as the details of the specific events are being worked out,
+during implementation of this proposal.
+
+##### Audit Log Timestamps
+
+The date format used in the audit logs should be something which can be easily parsed by fluentbit, to make it easy for
+users to configure fluentbit. We could easily document this to provide instructions on how to configure a custom
+fluentbit parser for Pinniped audit logs. We should probably
+avoid [fluentbit's default json parser's](https://github.com/fluent/fluent-bit/blob/845b6ae8576077fd512dbe64fb8e16ff4b15abdb/conf/parsers.conf#L35-L39)
+date format, which assumes dates will be in an ugly format and also lacks sub-second precision
+(e.g. `08/Apr/2022:19:24:01 +0000`).
+
+fluentbit uses [strptime](https://linux.die.net/man/3/strptime)
+with [an extension for fractional seconds](https://docs.fluentbit.io/manual/pipeline/parsers/configuring-parser#time-resolution-and-fractional-seconds)
+to parse timestamps.
+
+It would be desirable for a timestamp to:
+
+1. Be human-readable (e.g. not seconds since an epoch)
+2. Be easily parsable by log parsers, especially fluentbit
+3. Be expressed in UTC time
+4. Use at least millisecond precision
+5. Use the consistent JSON key name `time`
```

Comment:
I like the name `time` because it is an English word, and is consistent with the default key name used by fluentbit (and used by syslog too).

It doesn't really matter though, because the name is configurable in fluentbit parsers, so we could use `ts` if we have a reason to prefer it.

Reply:
I do not have a preference on the name itself, but it would be nice to be consistent.